crates/rbuilder/src/provider/ipc_state_provider.rs

```diff
@@ -0,0 +1,580 @@
+            // to IPC, then rbuilder cannot work
+            .expect("can't connect to IPC");
+
+        Self {
+            ipc_provider,
+            code_cache: Arc::new(DashMap::new()),
+            state_provider_by_hash: Arc::new(Cache::new(DEFAULT_STATE_CACHE_SIZE)),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
+#[serde(default, deny_unknown_fields)]
+pub struct IpcProviderConfig {
+    pub(crate) request_timeout_ms: u64,
+    pub(crate) ipc_path: PathBuf,
+    pub(crate) mempool_server_url: String,
+}
+
+impl Default for IpcProviderConfig {
+    fn default() -> Self {
+        Self {
+            request_timeout_ms: DEFAULT_IPC_REQUEST_TIMEOUT_MS,
+            mempool_server_url: String::new(),
+            ipc_path: PathBuf::new(),
+        }
+    }
+}
+
+impl StateProviderFactory for IpcStateProviderFactory {
+    /// Gets state for the latest block
+    fn latest(&self) -> ProviderResult<StateProviderBox> {
+        let state = IpcStateProvider::into_boxed(
+            self.ipc_provider.clone(),
+            BlockNumberOrTag::Latest.into(),
+            self.code_cache.clone(),
+        );
+
+        Ok(state)
+    }
+
+    /// Gets state at the block number
+    // We are not caching state provider by block number to avoid any issues with reorgs
+    // The calls to  history_by_block_number are rare and we shouldn't be loosing perf by not
+    // leveraging caching here
+    fn history_by_block_number(&self, block: BlockNumber) -> ProviderResult<StateProviderBox> {
+        let state = IpcStateProvider::into_boxed(
+            self.ipc_provider.clone(),
+            block.into(),
+            self.code_cache.clone(),
+        );
+
+        Ok(state)
+    }
+
+    /// Gets state at the block hash
+    fn history_by_block_hash(&self, block: BlockHash) -> ProviderResult<StateProviderBox> {
+        if let Some(state) = self.state_provider_by_hash.get(&block) {
+            return Ok(Box::new(state));
+        }
+
+        let state = IpcStateProvider::into_boxed(
+            self.ipc_provider.clone(),
+            block.into(),
+            self.code_cache.clone(),
+        );
+
+        self.state_provider_by_hash.insert(block, *state.clone());
+        Ok(state)
+    }
+
+    /// Gets block header given block hash
+    fn header(&self, block_hash: &BlockHash) -> ProviderResult<Option<Header>> {
+        let span =
+            trace_span!("header", id = rand::random::<u64>(), block_hash = %block_hash.to_string());
+        let _guard = span.enter();
+        trace!("header: get");
+
+        let header = self
+            .ipc_provider
+            .call::<_, Option<<alloy_network::Ethereum as alloy_network::Network>::BlockResponse>>(
+                "eth_getBlockByHash",
+                (block_hash, false),
+            )
+            .map_err(ipc_to_provider_error)?
+            .map(|b| b.header.inner);
```

Comment:
Rpc calls seems to be generic over parameter and return, and with the same error handling mechanism. 
Consider extracting it in a function to be able to call `let res: U = self.rpc_call("eth_method", params)?;` with a small wrapper like:
```rust
fn rpc_call<T, U>(&self, rpc_method: &str, params: T) -> ProviderResult<Option<U>> {
    self.ipc_provider
        .call::<T, Option<U>>(method, params)
        .map_err(ipc_to_provider_error)
}
```

Reply:
It could also encapsulates the tracing/span directly

Reply:
Great catch, done!